db/sql/cached_transaction.go

```diff
@@ -0,0 +1,189 @@
+// limitations under the License.
+
+package sql
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cloudwan/gohan/db/pagination"
+	"github.com/cloudwan/gohan/db/transaction"
+	"github.com/cloudwan/gohan/schema"
+	// Import mysql lib
+	_ "github.com/go-sql-driver/mysql"
+	// Import go-sqlite3 lib
+	_ "github.com/mattn/go-sqlite3"
+	// Import go-fakedb lib
+	"github.com/cloudwan/gohan/metrics"
+	"github.com/mitchellh/hashstructure"
+	_ "github.com/nati/go-fakedb"
+)
+
+type CachedState struct {
+	list     []*schema.Resource
+	total    uint64
+	isLocked bool
+}
+
+type CachedTransaction struct {
+	TxInterface
+	QueryCache map[string]CachedState
+}
+
+func MakeCachedTransaction(transx TxInterface) TxInterface {
+	cachedTransaction := &CachedTransaction{transx, nil}
+	cachedTransaction.ClearCache()
+	return cachedTransaction
+}
+
+func (tx *CachedTransaction) Create(resource *schema.Resource) error {
+	return tx.CreateContext(context.Background(), resource)
+}
+
+func (tx *CachedTransaction) CreateContext(ctx context.Context, resource *schema.Resource) error {
+	tx.ClearCache()
+	return tx.TxInterface.CreateContext(ctx, resource)
+}
+
+func (tx *CachedTransaction) Update(resource *schema.Resource) error {
+	return tx.UpdateContext(context.Background(), resource)
+}
+
+func (tx *CachedTransaction) UpdateContext(ctx context.Context, resource *schema.Resource) error {
+	tx.ClearCache()
+	return tx.TxInterface.UpdateContext(context.Background(), resource)
+}
+
+func (tx *CachedTransaction) StateUpdate(resource *schema.Resource, state *transaction.ResourceState) error {
+	return tx.StateUpdateContext(context.Background(), resource, state)
+}
+
+func (tx *CachedTransaction) StateUpdateContext(ctx context.Context, resource *schema.Resource, state *transaction.ResourceState) error {
+	tx.ClearCache()
+	return tx.TxInterface.StateUpdateContext(context.Background(), resource, state)
+}
+
+func (tx *CachedTransaction) Delete(s *schema.Schema, resourceID interface{}) error {
+	return tx.DeleteContext(context.Background(), s, resourceID)
+}
+
+func (tx *CachedTransaction) DeleteContext(ctx context.Context, s *schema.Schema, resourceID interface{}) error {
+	tx.ClearCache()
+	return tx.TxInterface.DeleteContext(context.Background(), s, resourceID)
+}
+
+func (tx *CachedTransaction) List(s *schema.Schema, filter transaction.Filter, options *transaction.ViewOptions, pg *pagination.Paginator) (list []*schema.Resource, total uint64, err error) {
+	return tx.ListContext(context.Background(), s, filter, options, pg)
+}
+
+//List resources in the db
+func (tx *CachedTransaction) ListContext(ctx context.Context, s *schema.Schema, filter transaction.Filter, options *transaction.ViewOptions, pg *pagination.Paginator) (list []*schema.Resource, total uint64, err error) {
+	sc := listContextHelper(s, filter, options, pg)
+
+	list, total, wasCached, _ := tx.getCached(s.ID, sc)
+
+	if !wasCached {
+		metrics.UpdateCounter(1, "tx.cache.miss")
```

Comment:
sorry not to be explicit about it before, but including the schema name in the metric path would be nice (eg.  https://github.com/cloudwan/gohan/blob/master/db/sql/sql.go#L590). note that it's ok to pass format+args to UpdateCounter (it's a nice, but premature optimization: if the metrics are turned off, we don't pay for building the string)

Reply:
Ok, will add it